Public/Src/Cache/ContentStore/Distributed/NuCache/RocksDbContentLocationDatabase.cs

```diff
@@ -765,55 +771,70 @@ private long DeserializeMetadataLastAccessTimeUtc(byte[] data)
             return DeserializeCore(data, reader => MetadataEntry.DeserializeLastAccessTimeUtc(reader));
         }
 
+        private class SortMetadataByFileTimeUtcComparer : IComparer<(long fileTimeUtc, byte[] strongFingerprint)>
+        {
+            public int Compare((long fileTimeUtc, byte[] strongFingerprint) x, (long fileTimeUtc, byte[] strongFingerprint) y)
+            {
+                return x.fileTimeUtc.CompareTo(y.fileTimeUtc);
+            }
+        }
+
         /// <inheritdoc />
         protected override BoolResult GarbageCollectMetadataCore(OperationContext context)
         {
             return _keyValueStore.Use(store => {
-                var cutoffTimeUtc = (Clock.UtcNow - _configuration.MetadataGarbageCollectionProtectionTime).ToFileTimeUtc();
-                int removedEntries = 0;
-                int scannedEntries = 0;
-                ulong dbSizeInBytesAfterGc = 0;
-                ulong dbRemovedBytesDuringGc = 0;
-
-                // TODO(jubayard): we may want to use a single WriteBatch here if this ends up affecting perf
-                foreach (var kvp in store.PrefixSearch((byte[])null, columnFamilyName: nameof(Columns.Metadata)))
+                // NOTE(jubayard): the most expensive part of this procedure is iterating over the whole database, so
+                // the less we take _while_ we do that, the better. An alternative to this procedure is to compute a
+                // quantile sketch and remove unneeded entries as we go.
+
+                var liveDbSizeInBytesBeforeGc = int.Parse(store.GetProperty(
+                    "rocksdb.estimate-live-data-size",
+                    columnFamilyName: nameof(Columns.Metadata)));
+
+                var scannedEntries = 0;
+                var removedEntries = 0;
+
+                var entries = new PriorityQueue<(long fileTimeUtc, byte[] strongFingerprint)>(
+                    capacity: _configuration.MetadataGarbageCollectionMaximumNumberOfEntries + 1,
+                    comparer: new SortMetadataByFileTimeUtcComparer());
+                foreach (var keyValuePair in store.PrefixSearch((byte[])null, nameof(Columns.Metadata)))
                 {
                     if (context.Token.IsCancellationRequested)
                     {
                         break;
                     }
 
-                    scannedEntries++;
-                    Counters[ContentLocationDatabaseCounters.GarbageCollectMetadataEntriesScanned].Increment();
-                    
-                    var lastAccessTimeUtc = DeserializeMetadataLastAccessTimeUtc(kvp.Value);
-                    if (lastAccessTimeUtc < cutoffTimeUtc)
+                    entries.Push((

```

Comment:
>entries [](start = 20, length = 7)

One optimization would be to compare to the top element first when the queue is over the limit. If its older, you don't need to add, you can just immediately remove.